Initial work on DuckDB

jdunkerley opened the pull request:
### Pull Request Description

- Initial set up for DuckDB
- Add connection and scaffolding
- Using Postgres dialect.

Next PR will add a dialect and bring into unit tests.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [ ] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [ ] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [ ] Unit tests have been written where possible.
- [ ] If meaningful changes were made to logic or tests affecting Enso Cloud integration in the libraries, 
      or the Snowflake database integration, a run of the [Extra Tests](https://github.com/enso-org/enso/actions/workflows/extra-nightly-tests.yml) has been scheduled.
  - If applicable, it is suggested to paste a link to a successful run of the Extra Tests.

jdunkerley committed:
Add DuckDB dependency.

jdunkerley committed:
Initial connectivity to DuckDB.

jdunkerley committed:
DuckDB doesn't support changing the database.

jdunkerley committed:
Add to export and make it work.

jdunkerley committed:
Initial reading from DuckDB with Postgres dialect.

jdunkerley committed:
Add read and make a more standard approach.

jdunkerley committed:
DuckDB license stuff

jdunkerley committed:
Add the rest of the methods to the connection.

jdunkerley committed:
Sort this out

jdunkerley committed:
Java format.

jdunkerley committed:
Lint and Docs.

jdunkerley committed:
Missing files.

jdunkerley committed:
MS API

JaroslavTulach commented:
- why don't we create `Standard.Duck` instead of putting new 3rd party library into `Standard.Database`?
- At least I believe it has been agreed that we want to move SQLite and Posgress out of `Standard.Database` _eventually_
- why not use proper architecture since the beginning in the case of DuckDB?
- or do we believe DuckDB is so essential it has to be in the core of all databases?

JaroslavTulach commented on distribution/lib/Standard/Microsoft/0.0.0-dev/docs/api/SQLServer_Connection.md:
```diff
@@ -24,6 +24,5 @@
     - tables self name_like:Standard.Base.Data.Text.Text= database:Standard.Base.Data.Text.Text= schema:Standard.Base.Data.Text.Text= types:Standard.Base.Any.Any= all_fields:Standard.Base.Any.Any= -> Standard.Base.Any.Any
     - to_js_object self -> Standard.Base.Any.Any
     - truncate_table self table_name:Standard.Base.Any.Any -> Standard.Base.Any.Any
-- make_table_name_selector connection:Standard.Base.Any.Any -> Standard.Base.Any.Any
```
- An API change in `Standard.Microsoft` when introducing a completely unrelated DB support!
- a sign of an unwanted change error
- or sign of previously low attention to API design
- probably the second, `make_table_name_selector` should have never been in the API, right?

jdunkerley replied:
Moving functionality that was added to the Microsoft library to the base library as generally reusable - reading from tables with schemas.

jdunkerley commented:
> * why don't we create `Standard.Duck` instead of putting new 3rd party library into `Standard.Database`?
> * At least I believe it has been agreed that we want to move SQLite and Posgress out of `Standard.Database` _eventually_
> * why not use proper architecture since the beginning in the case of DuckDB?
> * or do we believe DuckDB is so essential it has to be in the core of all databases?

At this point in testing it, DuckDB is expected to be very core to deliver some of the base functionality, like spatial and some reading and writing of files. If however it doesn't pan out to be useful, I will isolate to a separate library once have gone further along the path.

JaroslavTulach commented:
> deliver some of the base functionality, like spatial

The fact that DuckDB is believed to be essential from _product point of view_ doesn't mean we need to compromise the architecture. It'd be better if none of the core Table&Database concepts (like spatial) were tight to particular implementation.

JaroslavTulach commented:
> * why don't we create `Standard.Duck` instead of putting new 3rd party library into `Standard.Database`?

- Creating new `Standard.Duck` library from scratch would allow us to more easy validate the #13172 approach
- We know we can run #14037 in dual JVM mode
- but we still haven't turned that mode on in production as #13804 requests
   - why? because we are afraid of **regressing**
- by creating new `Standard.Duck` we automatically start executing it in _dual JVM mode_
- that will help us _design for dual JVM mode_ from scratch and help us fine tune everything since beginning
- **without risking any regressions** associated with changing the execution mode later

jdunkerley commented:
At least how I envision using DuckDB this will be in the NI space for a period of time.

I will put it as its own library and add to the standard list of import - assuming can do quickly as need to move on this.

jdunkerley committed:
Still got to sort CI jobs.

jdunkerley committed:
Prettier.

GregoryTravis commented on distribution/lib/Standard/Database/0.0.0-dev/src/Connection/Parquet_Format.enso:
```diff
@@ -0,0 +1,9 @@
+from Standard.Base import all
```
What does this file of just imports do?

jdunkerley committed:
Legal Review again

Doh

jdunkerley committed:
Initial DuckDB tests.

jdunkerley committed:
Remove broken test.

jdunkerley committed:
lock file

jdunkerley committed:
Legal review again

jdunkerley committed:
Docs and empty file.

jdunkerley committed:
Bazel lock again

jdunkerley committed:
Fix Bazel issue I hope!

jdunkerley committed:
Add a placeholder Java function for the DuckDB jar.

Akirathan commented on build.sbt:
```diff
@@ -6045,6 +6058,43 @@ lazy val `std-saas` = project
+    },
+    clean := Def.task {
+      val _ = clean.value
+      IO.delete(`std-aws-polyglot-root`)
```
copy & paste error:

```suggestion
      IO.delete(`std-duckdb-polyglot-root`)
```

Akirathan replied:
Note that, if left, it will, once again, cause issues with cleaning and caching builds

Akirathan reviewed (approved):
I suppose adding DuckDB to NI is OK. I am curious if you will be able to pass the [NI size check](https://github.com/enso-org/enso/blob/1d82231ceab99c3e9a8057d45239ad689a5efc03/project/NativeImage.scala#L362) though.

jdunkerley committed:
Java format.

jdunkerley committed:
Update build.sbt

Co-authored-by: Pavel Marek <pavel.marek@enso.org>

jdunkerley committed:
Doc update.

JaroslavTulach commented on build.sbt:
```diff
@@ -3931,6 +3937,9 @@ lazy val `engine-runner` = project
             .listFiles("*.jar")
             .map(_.getAbsolutePath()) ++
           `std-tableau-polyglot-root`
+            .listFiles("*.jar")
```
- I would really prefer if we could go without these three lines 
- it is an excellent opportunity to make _dual JVM mode_ work smoothly
- I promise to address any inquiries ASAP
- btw. try the experience execute:

```
sbt:enso> runEngineDistribution --jvm 
  --run test/DuckDB_Tests 
  --vm.D=polyglot.enso.classLoading=Standard.DuckDB:guest,enso_dev.DuckDB_Test:guest,hosted

4 tests succeeded.
0 tests failed.
0 tests skipped.
0 groups skipped.
```
- does it work? Yes, then let's try it. No, tell me what's failing!

jdunkerley replied:
Test code:
```
from Standard.Base import all
from Standard.Table import all
from Standard.Database import all
from Standard.DuckDB import all

main =
    connection5 = Database.connect DuckDB.In_Memory
    table4 = connection5.execute_query 'SELECT * FROM "/mnt/c/Users/jdunk/Downloads/flights-1m.parquet"' ..All_Rows write_operation=False
    any10 = table4.aggregate columns=[..Count, ..Maximum 'ARR_DELAY']
    any10.display
```

jdunkerley replied:
Data file: https://www.agentsfordata.com/sample-datasets/flights-1m.parquet

JaroslavTulach commented on distribution/lib/Standard/Database/0.0.0-dev/src/Connection/Connection.enso:
```diff
@@ -593,6 +593,20 @@ make_table_name_selector connection =
+## ---
+   private: true
+   ---
+make_table_name_with_schema_selector connection schema_black_list=[] =
```
- Can we put this method outside of public API?
- unless I am mistaken `Connection` is expose via re-export when `from Standard.Database import all`
- there is no point to put `## private: true` methods into types that are such prominently exposed
- why not create `src/Connecttion/Spi/Helpers.enso` and put the _service provider interface (SPI)_ methods there?

jdunkerley replied:
This is a widget creator used by all our different connection types.

I actually plan to restructure connectors into intersection types and that should make it better.

Static module methods are not exposed in the import all as the Connection type is exported not the module.

jdunkerley committed:
Use JDBC within DuckDB library rather than Database.

jdunkerley committed:
Fix for setScheam on DuckDB.

Commit:
Extract native libs from DuckDB jar

Commit:
update test: Native lib can have `.so.*` extension

hubertp committed:
Ensure duckdb is initialized at runtime

JaroslavTulach commented on build.sbt:
```diff
@@ -5531,6 +5531,24 @@ lazy val `sqlite-wrapper` = project
+      "libduckdb_java.so_osx_universal" -> PolyglotLib(MacOSAMD64),
+      "libduckdb_java.so_windows_amd64" -> PolyglotLib(WindowsAMD64),
+      "META-INF/**"                     -> CopyToOutputJar,
+      "org/**/*.class"                  -> CopyToOutputJar
```
I am getting:
```
sbt:enso> runEngineDistribution --jvm --run test.enso

Execution finished with an error: Can't load library: /enso/build/debug/libduckdb_java.so_linux_amd64
        at <java> java.base/java.lang.ClassLoader.loadLibrary(ClassLoader.java:2249)
        at <java> java.base/java.lang.Runtime.load0(Runtime.java:767)
        at <java> java.base/java.lang.System.load(System.java:1624)
        at <java> org.duckdb.DuckDBNative.<clinit>(DuckDBNative.java:52)
        at <java> org.duckdb.DuckDBConnection.newConnection(DuckDBConnection.java:62)
        at <java> org.duckdb.DuckDBDriver.connect(DuckDBDriver.java:114)
        at <java> java.sql/java.sql.DriverManager.getConnection(DriverManager.java:613)
        at <java> java.sql/java.sql.DriverManager.getConnection(DriverManager.java:160)
        at <java> org.enso.base.enso_cloud.EnsoSecretHelper.getJDBCConnection(EnsoSecretHelper.java:51)
```
where the [test.enso](https://github.com/enso-org/enso/pull/14114#discussion_r2413544958). We need to patch `DuckDBNative.java` (upstream as well, please) to use `System.loadLibrary`...

hubertp replied:
Yes, CI reports the same (it is masked by ClassNotFoundException)

hubertp replied:
Related https://github.com/duckdb/duckdb-java/issues/180

JaroslavTulach commented on .gitignore:
```diff
@@ -115,6 +115,9 @@ bench-report*.xml
 *.pdb
 *.so
 *.jar
+*.so_linux_amd64
```
- This is wrong idea. 
- We need `.so` extension otherwise `System.loadLibrary("duckdb")` will not find it
- but we need to modify [DuckDBNative](https://github.com/duckdb/duckdb-java/blob/2d2b38f544c4669ac9b020b57288fee53fe711c1/src/main/java/org/duckdb/DuckDBNative.java#L19)
- to call `System.loadLibrary` first before trying the equilibristic with extracting to `File.createTemp`

hubertp replied:
Logic in `DuckDBNative` appears to have been added to support local dev. I agree that logic is rather flawed if one wants to use the jar as a 3rd party dependency.

JaroslavTulach committed:
Make  Internal.DuckDB_Entity_Naming_Properties private

JaroslavTulach commented on distribution/lib/Standard/DuckDB/0.0.0-dev/src/DuckDB_Connection.enso:
```diff
@@ -0,0 +1,406 @@
+        jdbc_connection = JDBC_Connection.from_java sql_exception=SQLException java_connection
+
+        ## For initial work using DuckDB, we will use the Postgres dialect.
+        DuckDB_Connection.Value (Connection.new jdbc_connection Dialect.postgres DuckDB_Entity_Naming_Properties.new) make_new
```
- using `Dialect.postgres` doesn't allow us to use the same trick as `Standard.Microsoft` does
- there is [SQLServer_Type_Mapping.column_fetcher_factory](https://github.com/enso-org/enso/pull/14037/files#diff-87d75df7f1bd49d37fd7adbb1c4323a90ec6b08e030c458f9c07714fea024b41R121) which loads a class from _other JVM_
- if we did the same for `DuckDB`, then following worked:

```
sbt:enso> runEngineDistribution --jvm --run test.enso --vm.D=polyglot.enso.classLoading=Standard.DuckDB:guest,hosted
Execution finished with an error: Cannot convert '2006-01-01'(language: Enso Polyglot Bridge, type: java.sql.Date) to Java type 'java.sql.Date': Unsupported target type.
        at <java> com.oracle.truffle.polyglot.PolyglotObjectProxyHandler.invoke(PolyglotObjectProxyHandler.java:112)
        at <java> jdk.proxy3/jdk.proxy3.$Proxy46.getDate(Unknown Source)
        at <java> org.enso.database.JDBCUtils.getLocalDate(JDBCUtils.java:35)
        at <java> org.enso.database.fetchers.ColumnFetcherFactory$DefaultColumnFetcherFactory$5.getValue(ColumnFetcherFactory.java:72)
        at <java> org.enso.database.fetchers.BaseColumnFetcher.append(BaseColumnFetcher.java:51)
        at <java> org.enso.database.fetchers.ColumnFetcher.readResultSet(ColumnFetcher.java:51)
        at <enso> Result_Set.result_set_to_table.aggregator<arg-2>(Result_Set.enso:38:13-126)
        at <enso> Result_Set.result_set_to_table.aggregator(Result_Set.enso:37-38)
        at <enso> Java_Problems.with_problem_aggregator(Java_Problems.enso:116:14-25)
        at <enso> Result_Set.result_set_to_table(Result_Set.enso:35-39)
        at <enso> case_branch.statement_setter(Connection.enso:404:25-143)
```
- where [test.enso](https://github.com/enso-org/enso/pull/14114/files#r2413544958) is exactly the snippet being executed

jdunkerley replied:
Will be working on a new dialect for DuckDB next

hubertp committed:
Patch DuckDBNative and rename shared libraries

DuckDBNative needed to be patched so that it can be loaded as a shared
library by name in Native Image.
Additionally, it looks like `duckdb_java` didn't follow any standards
when it comes to naming the library name:
- always prefixed with `lib` (not acceptable on Windows)
- suffix includes OS and arch (should be one of `.so`, `.dll` or
  `.dylib`)

hubertp committed:
Add reachability metadata to fix JNI issues

JaroslavTulach commented on lib/java/duckdb-wrapper/src/main/java/org/duckdb/DuckDBNative.java:
```diff
@@ -0,0 +1,178 @@
+    static {
+        var libName = "duckdb_java";
+        try {
+            System.loadLibrary(libName);
```
I'd try to push this `System.loadLibrary("duckdb_java")` upstream to see how open the project is to 3rd party contributions.

JaroslavTulach commented on build.sbt:
```diff
@@ -5518,6 +5532,31 @@ lazy val `sqlite-wrapper` = project
+    inputJar := "org.duckdb" % "duckdb_jdbc" % duckdbVersion,
+    version := "0.1",
+    jarExtractor := JarExtractor(
+      "libduckdb_java.so_linux_amd64"   -> PolyglotLib(LinuxAMD64),
```
This creates a file with wrong (e.g. not `.so`) extension:
```
enso$ ls -1 built-distribution/enso-engine-*/enso-*/lib/Standard/DuckDB/*/polyglot/lib/amd64/linux/
libduckdb_java.so_linux_amd64
```
how does that get loaded into our JVM? `System.loadLibrary` is unlikely going to recognize this wrong extension!

In my opinion we should rename the file when extracting it and create
```
enso$ ls -1 built-distribution/enso-engine-*/enso-*/lib/Standard/DuckDB/*/polyglot/lib/amd64/linux/
libduckdb_java.so
```
Btw. after using `Standard.DuckDB` there is a new file in `/tmp/libduckdb_java15317788661761416119.so` - e.g. maybe we still let Duck DB extract during runtime!? And the library gets loaded in from the temporary file?

hubertp replied:
Are you sure this is not some cruft from your old builds?
```
> ls built-distribution/enso-engine-0.0.0-dev-linux-amd64/enso-0.0.0-dev/lib/Standard/DuckDB/0.0.0-dev/polyglot/lib/amd64/linux/
libduckdb_java.so
```

hubertp replied:
>  In my opinion we should rename the file when extracting it and create

We do. See changes in `JarExtractor`.

JaroslavTulach replied:
> Are you sure this is not some cruft from your old builds?

Probably. I performed `rm -rf built-distribution`, but that's probably is not enough. The `duckdb-wrapper` may probably need to be cleaned. I'll verify with full clean build today.

jdunkerley committed:
Notices

jdunkerley committed:
Prettier.
